### ISSUE TITLE
Improve Prompt Templates

### ISSUE THREAD

**Issue by langchain4j** [langchain4j/langchain4j#2118]
Consider adding a few improvements to current prompt template features:
- nested properties
- loops
- conditional expressions

Consider using existing (lightweight) templating engine (can be different ones for vanilla LC4j and SB integration) for this.

**Comment by langchain4j** [comment]
cc @glaforge (you were interested in this)

**Comment by glaforge** [comment]
So the idea would be to bring back templating like mustache or jinja (often used in the Python ecosystem) as a dedicated module maybe? Or you want to extend the current variable placeholder mechanism?

**Comment by langchain4j** [comment]
I guess we need to consider and evaluate all options. Ofc it would be great to have more features by default, but if it will require importing heawyweight library, then it is a no-go. Having a separate out-of-the-box module that enables better templating sounds good, but I would investigate first if there is lightweight option to include by default.

**Comment by glaforge** [comment]
I was looking at Jinja, because it's what's used by LangChain, so it could help porting nice prompts from LangChain to LangChain4j, however the jinjava library has [tons of dependencies](https://github.com/HubSpot/jinjava/blob/master/pom.xml). On the other hand, the jmustache library has [zero dependencies](https://github.com/samskivert/jmustache/blob/master/pom.xml) at all.
But there are other options as well...

My fear is that we turn the existing placeholder value replacement into something a bit too complex, and we invent yet another template engine :-)

**Comment by langchain4j** [comment]
Let's not reinvent the wheel and search for a lightweight solution that we can include with LC4j by default. If there is no such option, we can always implement a separate module that will optionally enable advanced templating features with one of the popular temlating engines (even if heavyweight)

**Comment by kpavlov** [comment]
This is an [example](https://github.com/kpavlov/langchain4j-kotlin/pull/27), how to introduce external template source and rendering mechanism it by extending LC4J. Here is the [documentation](https://github.com/kpavlov/langchain4j-kotlin/blob/main/docs/PromptTemplates.md) and [test](https://github.com/kpavlov/langchain4j-kotlin/blob/a7f671947d49d26f308e965c58d3b4234a393c43/langchain4j-kotlin/src/test/kotlin/me/kpavlov/langchain4j/kotlin/service/ServiceWithPromptTemplatesTest.kt#L21-L20). 

**TL;DR:**
```kotlin
interface Assistant {
    @UserMessage(
        // "Hello, {{userName}}! {{message}}"
        "prompts/ServiceWithTemplatesTest/default-user-prompt.mustache", 
    )
    fun askQuestion(
        @UserName userName: String,
        @V("message") question: String,
    ): String
}

val assistant = AiServices
      .builder(Assistant::class)
      .systemMessageProvider(
          TemplateSystemMessageProvider(
              //  "You are helpful assistant using chatMemoryID={{chatMemoryID}}"
              "prompts/ServiceWithTemplatesTest/default-system-prompt.mustache",
          ),
      ).chatLanguageModel(model)
      .build()

val response = assistant.askQuestion(userName = "My friend", question = "How are you?")
```


Regarding jmustache, it is important to make sure, that code execution via template is not allowed, to prevent prompt injection attack vector.
Simple placeholder replacement is [very easy to achieve](https://github.com/kpavlov/langchain4j-kotlin/blob/main/langchain4j-kotlin/src/main/kotlin/me/kpavlov/langchain4j/kotlin/prompt/SimpleTemplateRenderer.kt) with no dependencies.

**Comment by kpavlov** [comment]
#2194 - First step: introduced TemplateRenderingEngine interface and moved template rendering from DefaultPromptTemplateFactory. [This logic ](https://github.com/kpavlov/langchain4j/blob/5f234212d1a0c4f8dfe32fcdaa527d871a1b80f5/langchain4j-core/src/main/java/dev/langchain4j/model/input/PromptTemplate.java#L112-L117)should move to DefaultReneringEngine in a follow-up PRs.
Using factories/ServiceLoader to load TemplateRenderingEngine is out of scope of this PR and should be done separately.
